### ISSUE TITLE
Simplified reporting for known types?

### ISSUE THREAD

**Issue by karlkfi** [google/go-cmp#195]
So, I'm trying to compare two multi-lined strings with `cmp.DIff` and it returns a diff that looks pretty bulky and kind of confusing given the string is not part of another object.

Example:
```
  strings.Join({
  	"{",
- 	` "accountType": "self",`,
  	` "name": "account",`,
  	` "environment": "self",`,
  	` "providerVersion": "v2",`,
  	` "type": "kubernetes",`,
  	` "skin": "v2",`,
+ 	` "accountType": "self",`,
  	` "cloudProvider": "kubernetes"`,
  	"}",
  }, "\n")
```

I get that the objectification of strings is there to make it obvious that it's a string inside an object, but is there a way to pass in an option or reporter to Diff so that it has simplified output?

Example:
```
  {
-  "accountType": "self",
   "name": "account",
   "environment": "self",
   "providerVersion": "v2",
   "type": "kubernetes",
   "skin": "v2",
+ "accountType": "self",
   "cloudProvider": "kubernetes"
  }
```

If this was possible, it would be a boon for comparison of json/yaml strings without having to Unmarshal them first (especially in cases where unmarshalling might fail). It would make Diff more useful for comparing file contents.

**Comment by dsnet** [comment]
> I get that the objectification of strings is there to make it obvious that it's a string inside an object,

You are correct that is the rationale for why. The reporter output prioritizes accurate reporting over readability (though both are important goals).

> there a way to pass in an option or reporter to Diff so that it has simplified output?

Philosophically, we're generally adverse to adding lots of options to control how things look. I'd prefer pursuing an approach that (by default) produces output that is never ambiguous, but decently readable.

**Comment by karlkfi** [comment]
I understand the reasoning. It just means I'll have to use a different tool to get the clean output I want for json/yaml file/string comparison without having to marshal them first (which can fail on bad syntax).

**Comment by dsnet** [comment]
How would something like the following look.

For a standalone string:
```diff
  string(
      """
      one
-     two
+     TWO
      three
      """
  )
```

For a struct string field:
```diff
  Foo{
    StringField: string(
        """
        ... // 23 identical lines
        zero
        one
-       two
+       TWO
        three
        four
        ... // 534 identical lines
        """
    ),
    IntField: 5,
  }
```

For a slice of strings:
```diff
  []string{
    "equal single line",
-   "inequal single line",
+   "INEQUAL single line",
    string(
        """
        {
-        "accountType": "self",
         "name": "account",
         "environment": "self",
         "providerVersion": "v2",
         "type": "kubernetes",
         "skin": "v2",
+        "accountType": "self",
         "cloudProvider": "kubernetes"
        }
        """
    ),
    "another string",
  }
```

Specifics of how it works:
* The syntax uses a custom triple-quoted string syntax to represent multi-line strings.
* We use the nicer syntax if the following conditions hold:
    * The string has a sufficient number of lines, is valid UTF-8, and consists of printable characters.
    * `"""` does not appear in the diff output
    * `...` does not appear in the diff output
    * For lines that change, the differences cannot be due to whitespace (otherwise it becomes really hard to tell the difference between things like spaces and non-breaking spaces or due to a trailing carriage return).
* Otherwise it falls back on the `strings.Join({...})` syntax, which is unambigious, but ugly.

**Comment by karlkfi** [comment]
I think the triple quotes is a great compromise for this, as long as it's still clear where the indentation starts. Whitespace changes are common when comparing formatted json/yaml for linting.

Is the string() wrapper necessary? Won't triple quotes indicate it's unambiguously a string?

**Comment by dsnet** [comment]
> it's still clear where the indentation starts.

The starting `"""` and ending `"""` will always be aligned on the same column, so you can tell any additional indentation after that point.

> Is the string() wrapper necessary? Won't triple quotes indicate it's unambiguously a string?

Not strictly. It was to give some degree of indention to help things read nicer. Consider the `[]string` example without it:
```diff
  []string{
    "equal single line",
-   "inequal single line",
+   "INEQUAL single line",
    """
    {
-    "accountType": "self",
     "name": "account",
     "environment": "self",
     "providerVersion": "v2",
     "type": "kubernetes",
     "skin": "v2",
+    "accountType": "self",
     "cloudProvider": "kubernetes"
    }
    """
    "equal\nmultiple\nlines",
    "another string",
+   "added string",
  }
```
Visually, it's hard to distinguish between the individual strings that belong to the `[]string` and the individual lines that belong to the triple-quoted string.

**Comment by dsnet** [comment]
See #212.
